job/base.go

```diff
@@ -308,23 +310,53 @@ func (b *Base) Scopes() []string {
 	return b.RequiredScopes
 }
 
-// SetShouldApplyScopesOnEnqueue overrides the default behavior of scopes so
-// that they apply when the job is inserted into the queue rather than when the
-// job is dispatched.
-func (b *Base) SetShouldApplyScopesOnEnqueue(val bool) {
+// SetEnqueueScopes overrides the default behavior for the given scopes so that
+// they apply when the job is inserted into the queue rather than when the job
+// is dispatched. This filter will be ignored if EnqueueAllScopes is true.
+func (b *Base) SetEnqueueScopes(scopes ...string) {
 	b.mutex.Lock()
 	defer b.mutex.Unlock()
 
-	b.applyScopesOnEnqueue = val
+	b.enqueueScopes = scopes
 }
 
-// ShouldApplyShouldScopesOnEnqueue returns whether the job's scopes are applied on
-// enqueue. If false, the scopes are applied when the job is dispatched.
-func (b *Base) ShouldApplyScopesOnEnqueue() bool {
+// EnqueueScopes returns the subset of the job's scopes that will be applied on
+// enqueue, if any.
+func (b *Base) EnqueueScopes() []string {
 	b.mutex.RLock()
 	defer b.mutex.RUnlock()
 
-	return b.applyScopesOnEnqueue
+	if b.enqueueAllScopes {
+		return b.RequiredScopes
+	}
+
+	scopes := utility.StringSliceIntersection(b.RequiredScopes, b.enqueueScopes)
+	if len(scopes) == 0 {
```

Comment:
Yes, this is just an extra safeguard, because nil vs. empty slice matters greatly for checking scopes in the DB. If this can return an empty slice when there's no scopes, the scope `$exists` check will no longer be valid.

Reply:
Fair. I think it's easier to understand from a human perspective if the enqueue scopes followed the same expected behavior as the regular scopes field, but it doesn't matter from the DB's perspective.